### ISSUE TITLE
zippedArchive will make bugs when application runs in multiple processes

### ISSUE THREAD

**Issue by helloqhx** [winstonjs/winston-daily-rotate-file#174]
if use the process manage tool like `pm2`, to control your app, the `zippedArchive` will lead to bug, because multiple processes will try to zip the log file,

the source code:

```if (options.zippedArchive) {    
          this.logStream.on('rotate', function (oldFile) {
               var gzip = zlib.createGzip();
               var inp = fs.createReadStream(oldFile);
                var out = fs.createWriteStream(oldFile + '.gz');
                inp.pipe(gzip).pipe(out).on('finish', function () {
                    fs.unlinkSync(oldFile);
                });
            });
        }
```

**Comment by mehdibeldji** [comment]
I have the same problem 👍 It's possible to fix it in a new version please ? An error listener on `fs.createReadStream` is missing. Actually, it generate an uncaught exception 😞

**Comment by mattberther** [comment]
I am not familiar enough with `pm2` to understand the best way to make this code safe. What are some best practices to use?

**Comment by mehdibeldji** [comment]
I think the best way it's to use [the cluster library of node](https://nodejs.org/api/cluster.html) for know if the node process run in cluster mode or not. For that, you need to check if `cluster.worker` is defined and if `cluster.worker.isConnected() === true`
 
Firstly, we need to add an error listener on `fs.createReadStream`. If the node process doesn't run in cluster mode, we need to log an error, else it's unnecessary to log it.

**Comment by ankitcf** [comment]
I am ready to make the change but what should be the desired behaviour ? Should we silently eat up the error if unarchived file does not exist anymore ? 

```
        if (options.zippedArchive) {
            this.logStream.on('rotate', function (oldFile) {
                var gzip = zlib.createGzip();
                var inp = fs.createReadStream(oldFile)
                inp.on('open', function() {
                    var out = fs.createWriteStream(oldFile + '.gz');
                    inp.pipe(gzip).pipe(out).on('finish', function () {
                        fs.unlinkSync(oldFile);
                    });
                })
                inp.on('error', function(err) {});
            });
        }
```
works for me locally but I am not sure if we should be eating up this error.

**Comment by Arunachalam** [comment]
> I am ready to make the change but what should be the desired behaviour ? Should we silently eat up the error if unarchived file does not exist anymore ?
> 
> ```
>         if (options.zippedArchive) {
>             this.logStream.on('rotate', function (oldFile) {
>                 var gzip = zlib.createGzip();
>                 var inp = fs.createReadStream(oldFile)
>                 inp.on('open', function() {
>                     var out = fs.createWriteStream(oldFile + '.gz');
>                     inp.pipe(gzip).pipe(out).on('finish', function () {
>                         fs.unlinkSync(oldFile);
>                     });
>                 })
>                 inp.on('error', function(err) {});
>             });
>         }
> ```
> works for me locally but I am not sure if we should be eating up this error.

Hi,
We're also facing the same issue in the application, where will we put above codes or will we upgrade this module here to get the fixes?

**Comment by Arunachalam** [comment]
Where can we put below codes for this fix?

if (options.zippedArchive) {
            this.logStream.on('rotate', function (oldFile) {
                var gzip = zlib.createGzip();
                var inp = fs.createReadStream(oldFile)
                inp.on('open', function() {
                    var out = fs.createWriteStream(oldFile + '.gz');
                    inp.pipe(gzip).pipe(out).on('finish', function () {
                        fs.unlinkSync(oldFile);
                    });
                })
                inp.on('error', function(err) {});
            });
        }

Kindly request you to provide clear answer here.

**Comment by ankitcf** [comment]
That comment was not the intended to be a fix for anyone using winston-daily-rotate-file but rather a change in the library. It was a suggestion to the team to fix the bug. @mattberther any comments ?

**Comment by mehdibeldji** [comment]
I think the best solution is :
```
if (options.zippedArchive) {
  this.logStream.on('rotate', function (oldFile) {
      var gzip = zlib.createGzip();
      var inp = fs.createReadStream(oldFile);
      inp.on('error', function() {
          var zipExist = fs.createReadStream(oldFile + '.gz');
          zipExist.on('error', function() {
              throw new Error('File ' + oldFile + 'does not exist');
          });
      });
      var out = fs.createWriteStream(oldFile + '.gz');
      inp.pipe(gzip).pipe(out).on('finish', function () {
          fs.unlinkSync(oldFile);
      });
  });
}
```

If an error occured during readStream of older log file, we check if the `.gz`file exist. If it does not exist, we throw a new Error.

**Comment by Arunachalam** [comment]
> That comment was not the intended to be a fix for anyone using winston-daily-rotate-file but rather a change in the library. It was a suggestion to the team to fix the bug. @mattberther any comments ?

**Thank you for responding the details.**

> I think the best solution is :
> 
> ```
> if (options.zippedArchive) {
>   this.logStream.on('rotate', function (oldFile) {
>       var gzip = zlib.createGzip();
>       var inp = fs.createReadStream(oldFile);
>       inp.on('error', function() {
>           var zipExist = fs.createReadStream(oldFile + '.gz');
>           zipExist.on('error', function() {
>               throw new Error('File ' + oldFile + 'does not exist');
>           });
>       });
>       var out = fs.createWriteStream(oldFile + '.gz');
>       inp.pipe(gzip).pipe(out).on('finish', function () {
>           fs.unlinkSync(oldFile);
>       });
>   });
> }
> ```
> If an error occured during readStream of older log file, we check if the `.gz`file exist. If it does not exist, we throw a new Error.

Please let me know, once your team fixed and integrated this issue in the winston-daily-rotate-file npm module.

**Comment by mehdibeldji** [comment]
I Make the pull request : https://github.com/winstonjs/winston-daily-rotate-file/pull/185

**Comment by mattberther** [comment]
Merged #185 and pushed to npm as `winston-daily-rotate-file@3.3.4`.
